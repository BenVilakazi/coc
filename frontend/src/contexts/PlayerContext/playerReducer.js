function joinLobby(state) {
  return {
    ...state,
    gameState: 'pending-connection',
    message: {
      big: 'Connecting to Lobby',
      small: 'Please wait',
    },
    loading: [...state.loading, 'joining-lobby'],
  };
}

function submitCards(state) {
  return {
    ...state,
    gameState: 'submitting-cards',
    message: {
      big: 'Submitting your cards',
      small: 'Please wait',
    },
    loading: [...state.loading, 'submitting-cards'],
  };
}

function update(state, payload) {
  const { removeLoading, ...newData } = payload;
  return {
    ...state,
    ...newData,
    loading: removeLoading
      ? state.loading.filter((loadingVal) => loadingVal !== removeLoading)
      : state.loading,
  };
}

function errorDisconnect(state) {
  return {
    ...state,
    gameState: 'error',
    message: {
      big: 'AN ERROR OCCURRED',
    },
  };
}

function receiveWhiteCards(state, payload) {
  const { cards, selectCardCount } = payload;

  return {
    ...state,
    gameState: 'player-select',
    cards,
    selectCardCount,
  };
}

function lobbyClosed(state) {
  return {
    ...state,
    gameState: 'lobby-closed',
    message: {
      big: 'THE LOBBY HAS BEEN CLOSED',
      small: "You don't have to go home, but you can't stay here",
    },
  };
}

function noLobby(state) {
  return {
    ...state,
    gameState: 'error',
    message: {
      big: 'Lobby does not exist',
      small: 'We understand, spelling is pretty hard.',
    },
    loading: [],
  };
}

function removeDisconnectedPlayersCard(state, { playerId }) {
  if (state.submittedCards && state.submittedCards.length > 0) {
    const cardRemoved = state.submittedCards.filter(
      (card) => card.playerID !== playerId,
    );

    return {
      ...state,
      submittedCards: cardRemoved || state.submittedCards,
    };
  }
  return state;
}

function reducer(state, action) {
  const { type, payload } = action;
  switch (type) {
    case 'JOIN_LOBBY':
      return joinLobby(state);

    case 'UPDATE':
      return update(state, payload);

    case 'ERROR_DISCONNECT':
      return errorDisconnect(state);

    case 'SUBMIT_CARDS':
      return submitCards(state);

    case 'RECEIVE_WHITE_CARDS':
      return receiveWhiteCards(state, payload);

    case 'LOBBY_CLOSED':
      return lobbyClosed(state);

    case 'NO_LOBBY':
      return noLobby(state);

    case 'REMOVE_DISCONNECTED_PLAYERS_CARD':
      return removeDisconnectedPlayersCard(state, payload);

    default:
      return { ...state };
  }
}

export default reducer;
