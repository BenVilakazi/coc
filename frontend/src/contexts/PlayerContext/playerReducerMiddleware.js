import socketInstance from '../../socket/socket';

function joinLobby({ lobbyId, playerName }) {
  socketInstance.joinLobby(lobbyId, playerName);
}

function submitCards(payload) {
  socketInstance.sendMessage({ event: 'player-submit', payload });
}

function previewWinner(payload) {
  socketInstance.sendMessage({ event: 'preview-winner', payload });
}

function submitWinner(payload) {
  socketInstance.sendMessage({ event: 'winner-selected', payload });
}

export default async function playerReducerMiddleware(
  { type, payload },
  dispatch,
) {
  switch (type) {
    case `JOIN_LOBBY`:
      joinLobby(payload);
      break;

    case `SUBMIT_CARDS`:
      submitCards(payload);
      break;

    case 'PREVIEW_WINNER':
      previewWinner(payload);
      break;

    case 'SUBMIT_WINNER':
      submitWinner(payload);
      break;

    default:
      break;
  }
  dispatch({ type, payload });
}
