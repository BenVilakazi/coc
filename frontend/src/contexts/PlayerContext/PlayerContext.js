import React, { createContext, useEffect } from 'react';
import PropTypes from 'prop-types';
import reducer from './playerReducer';
import socketInstance from '../../socket/socket';
import useReducerMiddleware from '../useReducerMiddleware';
import playerReducerMiddleware from './playerReducerMiddleware';

// when the socket receives messaged from the server, it will fire off an event from this EventEmitter
// importing it allows us to "subscribe" to those events and update our state as needed
const { emitter } = socketInstance;

const initialState = {
  gameState: 'enter-code',
  cards: [],
  message: {
    big: '',
    small: '',
  },
  selectCardCount: 0,
  loading: [],
};

const propTypes = {
  children: PropTypes.node.isRequired,
};

// subscribe to the player context
export const PlayerContext = createContext();

export function PlayerProvider({ children }) {
  const [state, dispatch] = useReducerMiddleware(
    playerReducerMiddleware,
    reducer,
    initialState,
  );
  // this function allows us to parse any incoming messages from the event emitter
  // to make sure we know how to handle them.
  function handleMessage({ event, payload }) {
    switch (event) {
      case 'join-lobby':
        return dispatch({ type: 'JOIN_LOBBY', payload });

      case 'update':
        return dispatch({ type: 'UPDATE', payload });

      case 'error-disconnect':
        return dispatch({ type: 'ERROR_DISCONNECT', payload });

      case 'deal-white-cards':
        return dispatch({ type: 'RECEIVE_WHITE_CARDS', payload });

      case 'lobby-closed':
        return dispatch({ type: 'LOBBY_CLOSED' });

      case 'no-lobby':
        return dispatch({ type: 'NO_LOBBY', payload });

      case 'remove-disconnected-players-card':
        return dispatch({ type: 'REMOVE_DISCONNECTED_PLAYERS_CARD', payload });

      default:
        return undefined;
    }
  }

  useEffect(() => {
    // subscribe to the event emitter
    emitter.on('message', handleMessage);
    // cleanup the listener when dismounting
    return () => {
      emitter.off('message', handleMessage);
    };
  }, []);

  return (
    <PlayerContext.Provider value={{ state, dispatch }}>
      {children}
    </PlayerContext.Provider>
  );
}

PlayerProvider.propTypes = propTypes;
