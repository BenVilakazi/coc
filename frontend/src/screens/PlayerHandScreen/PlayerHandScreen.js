import React, { useState, useContext } from 'react';
import { PlayerContext } from '../../contexts/PlayerContext/PlayerContext';

import CardHandLayout from '../../layouts/CardHandLayout/CardHandLayout';
import PlayerHand from '../../components/PlayerHand/PlayerHand';

export default function PlayerHandScreen() {
  const { state, dispatch } = useContext(PlayerContext);

  const [selected, setSelected] = useState([]);

  const handleClear = () => setSelected([]);

  const handleSubmit = () => {
    dispatch({
      type: 'SUBMIT_CARDS',
      payload: { selectedCards: selected },
    });
  };

  // limits how many cards can be selected based on the 'cardsNeededForSelection'
  const updateSelected = (selectedCards) => {
    if (selectedCards.length <= state.selectCardCount) {
      setSelected(selectedCards);
    }
  };

  return (
    <>
      <CardHandLayout
        title={{
          top: 'PLAYER,',
          bottom: `SUBMIT ${state.selectCardCount} CARD${
            state.selectCardCount !== 1 ? 'S' : ''
          }`,
        }}
        onClear={handleClear}
        onSubmit={handleSubmit}
        numberSelected={selected.length}
      >
        <PlayerHand
          cards={state.cards}
          selectedCards={selected}
          onSelect={(selectedCards) => updateSelected(selectedCards)}
        />
      </CardHandLayout>
    </>
  );
}
