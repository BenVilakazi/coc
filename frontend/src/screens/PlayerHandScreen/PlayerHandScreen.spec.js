import React from 'react';
import { render, screen, act } from '@testing-library/react';
import renderer from 'react-test-renderer';
import userEvent from '@testing-library/user-event';
import PlayerHandScreen from './PlayerHandScreen';
import { PlayerContext } from '../../contexts/PlayerContext/PlayerContext';

describe('PlayerHandScreen', () => {
  describe('render', () => {
    const state = {
      cards: [
        'Card One',
        'Card Two',
        'Card Three',
        'Card Four',
        'Card Five',
        'Card Six',
        'Card Seven',
        'Card Eight',
        'Card Nine',
        'Card Ten',
      ],
    };

    const dispatch = jest.fn();

    it('should render', () => {
      render(
        <PlayerContext.Provider value={{ state, dispatch }}>
          <PlayerHandScreen />
        </PlayerContext.Provider>,
      );

      expect(screen.getByText('PLAYER,')).toBeInTheDocument();
    });

    it('should have the proper pluralization of the word "card" when selectCardCount is one', () => {
      const newState = { ...state, selectCardCount: 1 };

      render(
        <PlayerContext.Provider value={{ state: newState, dispatch }}>
          <PlayerHandScreen />
        </PlayerContext.Provider>,
      );

      expect(screen.getByText('SUBMIT 1 CARD')).toBeInTheDocument();
    });

    it('should have the proper pluralization of the word "card" when selectCardCount is not one', () => {
      const newState = { ...state, selectCardCount: 2 };

      render(
        <PlayerContext.Provider value={{ state: newState, dispatch }}>
          <PlayerHandScreen />
        </PlayerContext.Provider>,
      );

      expect(screen.getByText('SUBMIT 2 CARDS')).toBeInTheDocument();
    });
  });

  describe('dispatch', () => {
    const state = {
      cards: [
        'Card One',
        'Card Two',
        'Card Three',
        'Card Four',
        'Card Five',
        'Card Six',
        'Card Seven',
        'Card Eight',
        'Card Nine',
        'Card Ten',
      ],
      selectCardCount: 3,
    };

    const dispatch = jest.fn();

    it('dispatches SUBMIT_CARDS with the proper payload when Submit is clicked', () => {
      render(
        <PlayerContext.Provider value={{ state, dispatch }}>
          <PlayerHandScreen />
        </PlayerContext.Provider>,
      );

      const cardWrappers = screen.queryAllByTestId('card-wrapper');
      userEvent.click(cardWrappers[0]);
      userEvent.click(cardWrappers[1]);
      userEvent.click(cardWrappers[2]);
      userEvent.click(screen.getByText('SUBMIT'));

      expect(dispatch).toHaveBeenCalledTimes(1);
      expect(dispatch).toHaveBeenCalledWith({
        type: 'SUBMIT_CARDS',
        payload: { selectedCards: [0, 1, 2] },
      });
    });
  });

  describe('snapshot', () => {
    const state = {
      cards: [
        'Card One',
        'Card Two',
        'Card Three',
        'Card Four',
        'Card Five',
        'Card Six',
        'Card Seven',
        'Card Eight',
        'Card Nine',
        'Card Ten',
      ],
    };

    const dispatch = jest.fn();

    it('matches', () => {
      const tree = renderer
        .create(
          <PlayerContext.Provider value={{ state, dispatch }}>
            <PlayerHandScreen />
          </PlayerContext.Provider>,
        )
        .toJSON();

      expect(tree).toMatchSnapshot();
    });
  });

  describe('props', () => {
    it('passes the correct value for numberSelected', () => {
      const state = {
        cards: [
          'Card One',
          'Card Two',
          'Card Three',
          'Card Four',
          'Card Five',
          'Card Six',
          'Card Seven',
          'Card Eight',
          'Card Nine',
          'Card Ten',
        ],
        selectCardCount: 2,
      };
      const dispatch = jest.fn();

      render(
        <PlayerContext.Provider value={{ state, dispatch }}>
          <PlayerHandScreen />
        </PlayerContext.Provider>,
      );

      expect(screen.getByTestId('submit')).toHaveTextContent('0/2 SELECTED');

      act(() => {
        userEvent.click(screen.getByText('Card One'));
      });

      expect(screen.getByTestId('submit')).toHaveTextContent('1/2 SELECTED');
    });
  });
});
