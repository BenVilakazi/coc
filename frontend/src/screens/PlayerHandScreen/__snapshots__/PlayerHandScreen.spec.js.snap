// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlayerHandScreen snapshot matches 1`] = `
.c2 {
  display: block;
  border: 2px solid var(--primary-text-color);
  font-size: 12px;
  font-weight: bold;
  padding: 0.7em;
  background: var(--secondary-background-color);
  color: var(--secondary-text-color);
}

.c2:hover,
.c2:active,
.c2:focus {
  background: var(--accent-background-color);
  cursor: pointer;
}

.c3 {
  display: block;
  border: 2px solid var(--primary-text-color);
  font-size: 12px;
  font-weight: bold;
  padding: 0.7em;
  background: var(--secondary-background-color);
  color: var(--secondary-text-color);
  background: var(--primary-background-color);
  color: var(--primary-text-color);
}

.c3:hover,
.c3:active,
.c3:focus {
  background: var(--accent-background-color);
  cursor: pointer;
}

.c0 {
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  height: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c0 .card-container {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  max-width: 100%;
  margin: auto 0 auto 0;
  overflow: auto;
  padding-left: 20px;
}

.c0 .card-container > * {
  margin-right: 20px;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.c1 {
  width: 100%;
  min-height: 85px;
  background-color: var(--secondary-background-color);
  color: var(--secondary-text-color);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c1 .header-txt {
  margin: 10px 25px -6px -2px;
  line-height: 200%;
  -webkit-align-self: flex-end;
  -ms-flex-item-align: end;
  align-self: flex-end;
  font-size: 18px;
}

.c1 h1 span {
  display: block;
}

.c1 .button-container {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 .button {
  text-align: center;
  padding: 0 10px 0 10px;
  height: 30px;
  width: 90px;
  margin-right: 15px;
  border: none;
}

.c5 {
  position: relative;
  border: 1px solid var(--primary-text-color);
  width: 148px;
  height: 206px;
  background-color: var(--primary-background-color);
  color: var(--primary-text-color);
  border-radius: 7px;
  font-weight: 500;
  font-size: 18px;
  overflow-wrap: break-word;
  line-height: 21px;
  padding: 17px 15px 0 15px;
  box-shadow: 0px 4px 12px rgba(0,0,0,0.25);
}

.c5 .whiteCardText {
  max-height: 150px;
  padding-right: 0;
  overflow-y: auto;
}

.c5 .whiteCardText div {
  padding-bottom: 20px;
}

.c5 .stackedCards {
  position: absolute;
  bottom: 7px;
  left: 11px;
  width: 32px;
}

.c5 .backOfCard {
  font-weight: 600;
}

.c4 {
  background: none;
  border: none;
  text-align: left;
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: start;
  -webkit-justify-content: start;
  -ms-flex-pack: start;
  justify-content: start;
  -webkit-align-items: start;
  -webkit-box-align: start;
  -ms-flex-align: start;
  align-items: start;
  padding: 16px 0px;
  border-bottom: 5px solid;
  border-bottom-color: transparent;
}

.c4 .badge {
  z-index: 10;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  position: absolute;
  height: 27px;
  width: 27px;
  right: -10px;
  top: 5px;
  border-radius: 50%;
  font-size: 16px;
  background-color: var(--secondary-background-color);
  color: var(--secondary-text-color);
  font-weight: 600;
}

@media (min-height:640px) and (min-width:600px) {
  .c0 .card-container {
    -webkit-flex-wrap: wrap;
    -ms-flex-wrap: wrap;
    flex-wrap: wrap;
    -webkit-box-pack: center;
    -webkit-justify-content: center;
    -ms-flex-pack: center;
    justify-content: center;
    padding-top: 40px;
  }

  .c0 .card-container > * {
    margin-bottom: 40px;
  }
}

@media (min-height:640px) and (min-width:600px) {
  .c1 {
    min-height: 100px;
  }

  .c1 .header-txt {
    margin: 26px 70px -7px -2px;
    font-size: 1.3rem;
  }

  .c1 .button-container {
    margin-right: 40px;
  }
}

@media (max-width:396px) {
  .c1 .header-txt {
    margin: 10px;
    margin-left: 0;
    font-size: 12px;
    margin-bottom: -4px;
  }
}

@media (min-width:397px) and (max-width:500px) {
  .c1 .header-txt {
    margin: 10px;
    margin-left: 0;
    font-size: 18px;
    margin-bottom: -7px;
  }
}

@media (min-width:1025px) {
  .c5 {
    width: 267px;
    height: 363px;
    font-size: 36px;
    line-height: 42px;
    padding: 30px 27px 0 27px;
  }

  .c5 .whiteCardText {
    max-height: 280px;
    padding-right: 0;
  }

  .c5 .stackedCards {
    width: auto;
  }
}

@media (min-width:1025px) {
  .c4 .badge {
    top: 0;
    right: -15px;
    height: 40px;
    width: 40px;
    font-size: 30px;
  }
}

<div
  className="c0"
>
  <header
    className="c1 header-container"
  >
    <div
      className="header-txt"
    >
      <h1>
        PLAYER,
        <span>
          SUBMIT undefined CARDS
        </span>
      </h1>
    </div>
    <section
      className="button-container"
    >
      <button
        className="c2 button"
        data-testid="submit"
        disabled={true}
        onClick={[Function]}
      >
        0/undefined SELECTED
      </button>
      <button
        className="c3 button"
        onClick={[Function]}
      >
        CLEAR
      </button>
    </section>
  </header>
  <div
    className="card-container"
  >
    <button
      className="c4"
      data-testid="card-wrapper"
      onClick={[Function]}
    >
      <div
        className="c5"
        data-testid="white-card"
      >
        <div
          className="whiteCardText"
        >
          <div>
            Card One
          </div>
        </div>
        <div
          className="stackedCards"
          data-testid="stacked-cards"
        >
          <img
            alt="stacked cards"
            src="cards-stack-offset-light-icon.svg"
          />
        </div>
      </div>
    </button>
    <button
      className="c4"
      data-testid="card-wrapper"
      onClick={[Function]}
    >
      <div
        className="c5"
        data-testid="white-card"
      >
        <div
          className="whiteCardText"
        >
          <div>
            Card Two
          </div>
        </div>
        <div
          className="stackedCards"
          data-testid="stacked-cards"
        >
          <img
            alt="stacked cards"
            src="cards-stack-offset-light-icon.svg"
          />
        </div>
      </div>
    </button>
    <button
      className="c4"
      data-testid="card-wrapper"
      onClick={[Function]}
    >
      <div
        className="c5"
        data-testid="white-card"
      >
        <div
          className="whiteCardText"
        >
          <div>
            Card Three
          </div>
        </div>
        <div
          className="stackedCards"
          data-testid="stacked-cards"
        >
          <img
            alt="stacked cards"
            src="cards-stack-offset-light-icon.svg"
          />
        </div>
      </div>
    </button>
    <button
      className="c4"
      data-testid="card-wrapper"
      onClick={[Function]}
    >
      <div
        className="c5"
        data-testid="white-card"
      >
        <div
          className="whiteCardText"
        >
          <div>
            Card Four
          </div>
        </div>
        <div
          className="stackedCards"
          data-testid="stacked-cards"
        >
          <img
            alt="stacked cards"
            src="cards-stack-offset-light-icon.svg"
          />
        </div>
      </div>
    </button>
    <button
      className="c4"
      data-testid="card-wrapper"
      onClick={[Function]}
    >
      <div
        className="c5"
        data-testid="white-card"
      >
        <div
          className="whiteCardText"
        >
          <div>
            Card Five
          </div>
        </div>
        <div
          className="stackedCards"
          data-testid="stacked-cards"
        >
          <img
            alt="stacked cards"
            src="cards-stack-offset-light-icon.svg"
          />
        </div>
      </div>
    </button>
    <button
      className="c4"
      data-testid="card-wrapper"
      onClick={[Function]}
    >
      <div
        className="c5"
        data-testid="white-card"
      >
        <div
          className="whiteCardText"
        >
          <div>
            Card Six
          </div>
        </div>
        <div
          className="stackedCards"
          data-testid="stacked-cards"
        >
          <img
            alt="stacked cards"
            src="cards-stack-offset-light-icon.svg"
          />
        </div>
      </div>
    </button>
    <button
      className="c4"
      data-testid="card-wrapper"
      onClick={[Function]}
    >
      <div
        className="c5"
        data-testid="white-card"
      >
        <div
          className="whiteCardText"
        >
          <div>
            Card Seven
          </div>
        </div>
        <div
          className="stackedCards"
          data-testid="stacked-cards"
        >
          <img
            alt="stacked cards"
            src="cards-stack-offset-light-icon.svg"
          />
        </div>
      </div>
    </button>
    <button
      className="c4"
      data-testid="card-wrapper"
      onClick={[Function]}
    >
      <div
        className="c5"
        data-testid="white-card"
      >
        <div
          className="whiteCardText"
        >
          <div>
            Card Eight
          </div>
        </div>
        <div
          className="stackedCards"
          data-testid="stacked-cards"
        >
          <img
            alt="stacked cards"
            src="cards-stack-offset-light-icon.svg"
          />
        </div>
      </div>
    </button>
    <button
      className="c4"
      data-testid="card-wrapper"
      onClick={[Function]}
    >
      <div
        className="c5"
        data-testid="white-card"
      >
        <div
          className="whiteCardText"
        >
          <div>
            Card Nine
          </div>
        </div>
        <div
          className="stackedCards"
          data-testid="stacked-cards"
        >
          <img
            alt="stacked cards"
            src="cards-stack-offset-light-icon.svg"
          />
        </div>
      </div>
    </button>
    <button
      className="c4"
      data-testid="card-wrapper"
      onClick={[Function]}
    >
      <div
        className="c5"
        data-testid="white-card"
      >
        <div
          className="whiteCardText"
        >
          <div>
            Card Ten
          </div>
        </div>
        <div
          className="stackedCards"
          data-testid="stacked-cards"
        >
          <img
            alt="stacked cards"
            src="cards-stack-offset-light-icon.svg"
          />
        </div>
      </div>
    </button>
  </div>
</div>
`;
